solidity/random-beacon/contracts/libraries/Groups.sol

```diff
@@ -148,29 +127,186 @@ library Groups {
+        // Move expiredGroupOffset as long as there are some groups that should
+        // be marked as expired. It is possible that expired group offset will
+        // move out of the groups array by one position. It means that all groups
+        // are expired (it points to the first active group) and that place in
+        // groups array - currently empty - will be possibly filled later by
+        // a new group.
+        while (
+            self.expiredGroupOffset < self.groupsRegistry.length &&
+            groupLifetimeOf(
+                self,
+                self.groupsRegistry[self.expiredGroupOffset]
+            ) <
+            block.number
+        ) {
+            self.expiredGroupOffset++;
+            self.activeGroupsCount--;
+        }
 
-    /// @notice Gets the number of active groups. Candidate, expired and terminated
-    ///         groups are not counted as active.
-    function numberOfActiveGroups(Data storage self)
+        // Go through all activeTerminatedGroups and if some of the terminated
+        // groups are expired, remove them from activeTerminatedGroups collection
+        // and rearrange the array to preserve the original order.
+        // This is needed because we evaluate the shift of selected group index
+        // based on how many non-expired groups have been terminated. Hence it is
+        // important that a number of terminated groups matches the length of
+        // activeTerminatedGroups[].
+        for (uint256 i = 0; i < self.activeTerminatedGroups.length; i++) {
+            if (self.expiredGroupOffset > self.activeTerminatedGroups[i]) {
+                // When 'i'th group qualifies for expiration, we need to remove
+                // it from the activeTerminatedGroups array manually by rearranging
+                // the order starting from 'i'th group.
+                for (
+                    uint256 j = i;
+                    j < self.activeTerminatedGroups.length - 1;
+                    j++
+                ) {
+                    self.activeTerminatedGroups[j] = self
+                        .activeTerminatedGroups[j + 1];
+                }
+                // Resizing the array length by 1. The last element was copied
+                // over in the loop above to an index "second to last". This is
+                // why we can safely remove it from here.
+                self.activeTerminatedGroups.pop();
+                // At this point the length of activeTerminatedGroups[] was shrinked
+                // by 1. We need to adjust 'i'th counter by 1 as well, otherwise
+                // it will overflow this array and throw an error.
+                i--;
+            }
+        }
+    }
+
+    /// @notice Terminates group with the provided index. Reverts if the group
+    ///         is already terminated.
+    /// @param  groupId Index in the groupRegistry array.
+    function terminateGroup(Data storage self, uint64 groupId) internal {
```

Comment:
Can we add `terminateGroup` call in `RandomBeacon.reportRelayEntryTimeout` and then address TODOs in `RandomBeacon.Relay.test.ts` regarding group termination?

Reply:
I've already branched off of this PR on my local and planning on creating a new PR that handles reporting. My plan for this PR is to focus on introducing group statuses and later address the reporting part.